### PR TITLE
fix(md-data-table): Move `MdDataTable`'s onChange subscription to `DataTableSelectableRow`

### PR DESCRIPTION
**Pull request by mailiam**
Binding from children allows async tr population to handle `onChange` event properly.
Also check `this.selectable` at event callback.

fixes where (onSelectableChange) not triggering when using http or other async row population

### CODE REVIEW

File: ng2-material/components/data_table/data_table.ts [hunk 1]
```diff
@@ -80,17 +84,6 @@ export class MdDataTable implements AfterContentInit {
...
-        this._masterRow.onChange.subscribe(this.change.bind(this));
-      }
-
-      this._rows.toArray()
-        .map((tr: MdDataTableSelectableRow) => {
-          tr.onChange.subscribe(this.change.bind(this));
-        });
-    }
-  }
+  ngAfterContentInit() {}
```
Comment (Gregcop1) [inline ng2-material/components/data_table/data_table.ts]: If `ngAfterContentInit` is no more required, why keeping it ?

**Comment by Gregcop1** [comment]
Can you write some tests about async tr population ? If you've solved an issue, some tests should be written on it

File: ng2-material/components/data_table/data_table.ts (hunk 1 above)
Comment (mailiam) [inline ng2-material/components/data_table/data_table.ts]: I thought it was for later use as other original components also had empty `ngAfterContentInit`.
I might as well clean it up. :smile:

**Comment by justindujardin** [comment]
Yes, please add a test or two to verify this change.  It looks good!

File: ng2-material/components/data_table/data_table_selectable_tr.ts [hunk 2]
```diff
@@ -82,6 +82,8 @@ export class MdDataTableHeaderSelectableRow extends AbstractMdDataTableSelectabl
     this.table.onSelectableChange
       .map(event => event.allSelected)
       .subscribe(newActiveStatus => this.isActive = newActiveStatus);
+
+    this.onChange.subscribe(this.table.change.bind(this.table));
```
Comment (justindujardin) [inline ng2-material/components/data_table/data_table_selectable_tr.ts]: I think that for this subscrpt and the one above, we should store the subscription results and unsubscribe() when the `OnDestroy` lifecycle event happens.  @Gregcop1 can you confirm that the onSelectableChange subscript should be doing this as well?

**Comment by mailiam** [comment]
I'm working on the test. @justindujardin could you please elaborate 'a test or two'? I'm new to Ng2 testing.

File: ng2-material/components/data_table/data_table_selectable_tr.ts (hunk 2 above)
Comment (Gregcop1) [inline ng2-material/components/data_table/data_table_selectable_tr.ts]: I'm not an expert on this subject but I think too we should force the unsubscribe on destroy. I'm not sure it's done at the moment for subscription between checkbox and master checkbox.

**Comment by mailiam** [comment]
@justindujardin @Gregcop1 I have committed some tests.

**Comment by justindujardin** [comment]
@mailiam it looks like the removal of the synchronous flag on the selectable change event causes dehydrated detector errors and fails some tests.

https://travis-ci.org/justindujardin/ng2-material/builds/124398884

**Comment by mailiam** [comment]
No idea why launcher.sauce fails.
It passes on my travis-ci(https://travis-ci.org/mailiam/ng2-material/builds/127115204)

**Comment by zylajoel** [comment]
Any updates on this?

**Pull request closed without merging**